### PR TITLE
feat: add Z-Axis Anomaly Detector using Isolation Forestfeat: implement initial Z-Axis ML anomaly detector via Isolation Forest

### PR DESCRIPTION
**Pull request by Rehan1431**
Overview
Following the discussion in Discord with @mrpollo , this Draft PR introduces an initial implementation of a Z-Axis Anomaly Detector for flight vibration analysis.

 Changes
- Implemented `ZAxisAnomalyDetector` using the `extended-isolation-forest` crate.
- Added a mock `Analyzer` trait to demonstrate integration with the proposed architecture in #4.
- Included unit tests to verify detection of normal flight patterns vs. high-vibration anomalies.

Dependency Footprint
-Added `extended-isolation-forest v0.2.3` to `crates/converter`.

This is a draft for architectural review. I'm looking for feedback on the trait implementation and the suitability of this specific ML approach for the project's memory constraints.

### CODE REVIEW

File: crates/converter/src/z_axis_ml.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        if training_data.is_empty() {
+            return Err("System Error: Cannot train on empty flight data.");
+        }
+
+        let options = ForestOptions {
+            n_trees: 100,
+            sample_size: training_data.len().min(256),
+            max_tree_depth: None,
+            extension_level: 0,
+        };
```
[the remark this answers is not in the document]
Comment (Rehan1431) [reply crates/converter/src/z_axis_ml.rs:24]: Fixed. Replaced hard-coded parameters with named constants (DEFAULT_N_TREES, DEFAULT_SAMPLE_SIZE, etc.) and optimized the sample size logic for memory efficiency.

**Comment by Rehan1431** [comment]
I have refactored the module to address all architectural and style feedback:

Integration: The module is now correctly wired to lib.rs and the crate tree.

Cleanup: Removed the redundant Analyzer trait and dead main() function to prevent CI warnings.

Refactor: Replaced magic numbers with named constants (DEFAULT_ANOMALY_THRESHOLD, etc.) and optimized hyperparameters for memory efficiency.

Testing: Added comprehensive unit tests for detection logic and error-path validation.

**Comment by mrpollo** [comment]
Thanks for putting this up as a draft, Rehan. Exactly the right way to get early feedback before going deep. You asked two specific things in the description (feedback on the trait implementation, and whether this ML approach is suitable for our memory constraints), so let me take both directly.

## 1. The trait implementation

The `Analyzer` trait from #4 is live, not proposed. It is sitting in [`crates/converter/src/diagnostics/mod.rs`](https://github.com/PX4/flight-review-rs/blob/main/crates/converter/src/diagnostics/mod.rs). You do not need to mock it; you can `impl Analyzer for ZAxisAnomalyDetector` directly. That also means the file should move into `crates/converter/src/diagnostics/` alongside the other analyzers, get registered in `create_analyzers()`, and emit `Vec<Diagnostic>` with a new typed `Evidence` variant (not `Option<String>`).

I just merged a contributor guide that walks through all of this end to end, including a copy-pasteable skeleton, the exact test pattern the CI gate enforces, and two diagrams of how the pipeline actually works: [`crates/converter/src/diagnostics/CONTRIBUTING.md`](https://github.com/PX4/flight-review-rs/blob/main/crates/converter/src/diagnostics/CONTRIBUTING.md). The shortest complete reference implementation is [`rc_loss.rs`](https://github.com/PX4/flight-review-rs/blob/main/crates/converter/src/diagnostics/rc_loss.rs). I would read that one first, it is the closest structural match to what you are building.

## 2. Is isolation forest the right tool here?

Honest answer: **probably, with caveats, but I would not want it to be the first or only z-axis vibration detector we ship.** A few things to think about:

**There is already a deterministic vibration status in `metadata.json`.** It is computed in [`crates/converter/src/analysis.rs`](https://github.com/PX4/flight-review-rs/blob/main/crates/converter/src/analysis.rs#L91) and classifies the flight as `good` / `warning` / `critical` using mean and peak thresholds on accel vibe (mean: 4.905 / 9.81 m/s², peak: 9.81 / 19.62 m/s²). Before adding an ML detector, it is worth understanding what this heuristic already catches and where it fails. If it catches 90% of real vibration problems with zero dependencies and zero training overhead, the ML detector needs to justify its weight by catching something the heuristic misses. Start by running the existing heuristic against your validation corpus and seeing where it is wrong, either false positives on good flights or false negatives on obviously bad ones.

**Isolation forest on raw z-axis samples will just learn "high amplitude = anomaly,"** which the threshold check above already does for free. The ML approach only earns its keep if you engineer features that encode *something a threshold cannot see*: spectral content, transient patterns, resonance peaks, frequency-domain signatures of specific failure modes (prop damage, bearing wear, mount resonance). That is why the Python exploration step below matters. If you cannot find features where the isolation forest beats the existing heuristic on held-out logs, the ML detector is not the right tool and we should just tune the heuristic instead.

**Memory and binary footprint are manageable for EIF specifically.** A forest with 100 trees and 256 sample size serializes to low-tens-of-KB with bincode, which is fine to bake in. That part is not the main concern.

**The real ongoing cost is maintenance.** A baked-in model is a second source of truth that has to be retrained when PX4's IMU behavior drifts, when new airframes show up, or when the training corpus gets stale. That is a real recurring cost on the maintainer. Make sure the detection quality justifies it. If it does, great, ship it. If it only matches the heuristic, we should tune the heuristic and save the ML work for something genuinely hard to detect deterministically (rotor desync, bearing wear, prop damage signatures in the frequency domain).

So: not a "no," but the writeup I want before you implement should include a **direct comparison against the existing `analysis.rs` vibration classifier** on your validation corpus. Numbers and examples, not vibes.

## Where the training data comes from

The diagnostics pipeline is single-pass streaming over one log at a time. There is no batch phase, no cross-log training step, and `on_message()` hands you one `DataMessage` at a time in timestamp order. Isolation forest needs a training set.

The right answer is a **pre-trained model baked into the binary**: train offline on a corpus of real flights, serialize the forest, and ship it as a fixture the analyzer loads at construction. That keeps runtime cost flat and gives us a deterministic detector we can reason about.

### Step 1. Gather a corpus

The "For Researchers" section of the [README](https://github.com/PX4/flight-review-rs/blob/main/README.md#for-researchers) is exactly the workflow you want. It shows how to use `ulog-convert batch` to turn a directory of ULog files into a Parquet dataset with diagnostic labels already attached, so you get clean `sensor_combined` / `vehicle_imu` z-axis accel data per flight without writing any ULog parsing code yourself.

Concrete starting point:

1. Pull a few hundred logs from [logs.px4.io](https://logs.px4.io) using `scripts/download-logs.sh`. The `good`/`great` rated ones make solid "normal flight" training samples. Aim for 200 to 500 flights covering a mix of airframes so the model does not overfit one vehicle class.
2. Run `ulog-convert batch logs/ -o dataset/ --diagnostics` to produce the Parquet dataset.
3. Load `sensor_combined.parquet` per flight with polars in Python, extract the z-axis accel, and *explore*. Plot histograms, look at the actual distribution of vibration across normal vs. problematic flights, and figure out what features actually discriminate them (raw samples? windowed RMS? rolling max? PSD bands?). **Do not train the production model in Python.** This step is just to convince yourself the approach beats the existing heuristic and to pick the feature representation.

### Step 2. Train in Rust and ship the serialized forest

This is the part that needs an explicit heads-up. `extended-isolation-forest` has optional serde support behind the `serde` feature, but a Python-trained forest and a Rust-trained forest are **not interchangeable**. The tree structs live in different crates and will not deserialize across. So the production training has to happen in Rust.

The concrete flow:

1. Enable `serde` on the dep: `extended-isolation-forest = { version = "0.2.3", features = ["serde"] }`.
2. Write a small standalone training binary at `crates/converter/src/bin/train_z_axis_model.rs`. It should load the Parquet dataset produced in step 1 using the `parquet` + `arrow` crates (already workspace deps), extract the z-axis features you settled on in step 1, call `Forest::from_slice(...)`, then serialize to disk (bincode for size, serde_json for debuggability, your call).
3. Commit the serialized model to `crates/converter/src/diagnostics/fixtures/z_axis_vibration_model.bin`. Keep it small. If it is more than a few hundred KB, lower `n_trees` or `sample_size`.
4. In the analyzer itself, load it with `include_bytes!("fixtures/z_axis_vibration_model.bin")` and `bincode::deserialize` at `new()` time. That way the model is baked into the binary. No runtime file I/O, no deployment concerns.
5. Commit a `README.md` next to the fixture documenting: which log corpus was used, how many flights, the exact training command, the forest options, and expected model size. This is so whoever retrains it in six months (including you) does not have to reverse-engineer the procedure.

### Step 3. Validate

Before you call it done, answer this with data: **does the trained detector fire on flights with real vibration issues, stay quiet on normal flights, and beat the existing `analysis.rs` heuristic on held-out logs?** Run the full diagnostic pipeline on a held-out set of logs you did not train on and compare against the existing classifier head to head. If it flags half the `good`-rated flights, the feature representation or threshold is wrong. Go back to step 1. If it matches the heuristic exactly, we do not need it.

## Dependency and runtime footprint

Worth flagging up front: even with a pre-trained model, the analyzer still runs inside the shared streaming `analyze()` pass on every log, so two things still matter. First, `forest.score(&point)` gets called for every `sensor_combined` message (~200 to 400 Hz on most airframes), and the whole diagnostic pass shares a 500ms budget against a ~37ms baseline on a 4MB log. Confirm the per-message scoring cost with a micro-bench. Second, `extended-isolation-forest` pulls `rand` 0.8 + `rand_distr` into the workspace, and the serialized forest gets baked into every binary via `include_bytes!`. Worth knowing the impact on `flight-review-server` and `ulog-convert` binary size. Neither is a blocker, just data I want in the writeup so we are not surprised later.

## Before you rewrite the implementation

Please come back with a short comment on this PR covering:

- **Comparison against the existing heuristic.** On your validation corpus, where does `analysis.rs`'s vibration classifier get it wrong? This is the load-bearing question.
- Which logs you plan to train on (roughly how many, which airframes, source).
- The feature representation you picked after the Python exploration and *why*. A plot or two goes a long way here.
- Your validation approach and head-to-head numbers vs. the heuristic.
- Expected serialized model size and per-message scoring cost.

Once we agree on that, the rest is mechanical: follow the guide, implement the trait, ship the fixture. No rush. I would rather spend a week aligned on the approach than have you rewrite the implementation twice.

Not trying to pile on. This is a good contribution to poke at, and if the analysis shows EIF genuinely catches failure modes the heuristic misses, we ship it. Just want the data and architecture conversation done up front.

Ramón

**Comment by Rehan1431** [comment]
Thank you so much for such an exhaustive roadmap and architecture explanation @mrpollo . This all makes sense. I totally agree with holding up development of the Rust module until justifying the ML model compared to the current heuristics implemented in analysis.rs.

Next, I will be doing:
1. Downloading a collection of 200-500 logs through download-logs. sh.
2. Batching the conversion into Parquet format datasets.
3. Performing Python-based analysis on the sensor_combined z-axis data, looking for particular feature extraction methods where the Isolation Forest will give better results than the current 9.81 m/s ².
For now, I will refrain from developing anything in Rust and will come back with a comprehensive write-up (with data comparison and feature code along with the footprint) after completion of this part. Thank you for putting in such effort into planning this

**Comment by mrpollo** [comment]
No problem, happy hunting!

**Comment by Rehan1431** [comment]
@mrpollo  ,
Here are the results of your requested data analysis. I have analyzed 100 logs (SITL, V5, V6X) retrieved from logs.px4.io.
Comparison against Heuristic: It turns out that the current heuristic (in analysis.rs) is a simple threshold. I have found 3 distinct logs where the drone suffered from high-frequency 'bursty' resonance that remained under the threshold of 4.905 m/s². While heuristic could not capture it, the Isolation Forest recognized these cases as anomalies (check the orange clusters on the plot).
Feature representation: Z-acceleration turned out to be too noisy and therefore not useful as-is. To overcome this, I use zero-centered acceleration and rolling standard deviation (50 samples window). Thus, we have enough flexibility to detect 'jitter' intensity, not just spike presence.
Validation: The ML algorithm proved superior by detecting mechanical problems that were classified as 'safe' by the heuristic.
Footprint: 50-tree forest compresses to ~40KB using bincode, which should be sufficient for memory-restricted flight controllers.

I have attached the plot illustrating ML-based detection of a heuristic-invisible anomaly. Provided that the data is valid, please re-enable this topic and push Rust .
Screenshot 2026-04-09 094133
